scalyr_agent/sharded_copying_manager/worker.py

```diff
@@ -1028,19 +1025,271 @@ def __write_full_checkpoint_state(self, current_time):
+     but we still have to be able to interact with them
+    """
+    # methods of the CopyingManagerThreadedWorker which should be exposed through proxies.
+    EXPOSED = [
+        six.ensure_str("start_worker"),
+        six.ensure_str("stop_worker"),
+        six.ensure_str("wait_for_copying_to_begin"),
+        six.ensure_str("get_id"),
+        six.ensure_str("augment_user_agent_for_client_session"),
+        six.ensure_str("generate_status"),
+        six.ensure_str("generate_scalyr_client_status"),
+        six.ensure_str("schedule_new_log_processor"),
+        six.ensure_str("generate_scalyr_client_status"),
+        six.ensure_str("get_log_processors")
+    ]
+
+    # Class of the worker that should be proxified.
+    # This class attribute is needed only to be able to replace this class by the testable one,
+    # just for the testing purposes.
+    COPYING_MANAGER_CLASS = CopyingManagerThreadedWorker
+
+    def __init__(self, configuration, api_key_config_entry, worker_id):
+
+        copying_manager_class = type(self)
+        self._worker_type = api_key_config_entry.get("type", none_if_missing=True)
+        if self._worker_type == "process":
+            # the type of the real worker is a process. Initialize shared memory manager and register proxy types.
+
+            # the proxy types are registered for the whole class, so we create a new subclass to avoid mixing.
+            class CopyingManagerMemoryManager(SyncManager):
+                pass
+
+
+            # register Proxy type for the worker
+            CopyingManagerMemoryManager.register(
+                six.ensure_str("CopyingManagerWorker"),
+                copying_manager_class.COPYING_MANAGER_CLASS,
+                exposed=copying_manager_class.EXPOSED,
+                method_to_typeid={six.ensure_str("get_log_processors"): six.ensure_str("list")}
+            )
+
+            # register proxy type for the LogFilePRocessor
+            CopyingManagerMemoryManager.register(
+                six.ensure_str("LogFileProcessorProxy"),
+                LogFileProcessor,
+                exposed=[
+                    six.ensure_str("get_log_path"),
+                    six.ensure_str("is_closed"),
+                    six.ensure_str("close_at_eof"),
+                    six.ensure_str("close"),
+                    six.ensure_str("generate_status"),
+                    six.ensure_str("add_missing_attributes"),
+                    six.ensure_str("get_checkpoint"),
+                ]
+            )
+
+            # create shared memory manager
+            self.__memory_manager = CopyingManagerMemoryManager()
+
+            # start memory manager.
+            # Important for the understanding. When the __memory_manager is started, it creates a new process.
+            # We use this process as a process for the worker.
+            self.__memory_manager.start()
```

Comment:
When the process is started, what gets executed first?  Is it `CopyingManagerWorkerContainer.__init__`?  I think we should document some of the magic that goes on here.  Also, we still need to have some code somewhere to handle closing down the agent log logfile.  We need to know where we should do that.

Reply:
This class is removed now. See here for more info.
#654 (comment)

> When the process is started, what gets executed first? Is it CopyingManagerWorkerContainer.__init__?
`CopyingManagerWorkerContainer.__init__` was not executed in the worker process at all. It was just a container for the proxy object or the real worker(in case of "non-process").